server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -594,6 +620,58 @@ private static ObjectMapper truncateObjectMapper(MergeReason reason, MapperMerge
         }
     }
 
+    /**
+     * Returns all FieldMappers this ObjectMapper or its children hold.
+     * The name of the FieldMappers will be updated to reflect the hierarchy.
+     *
+     * @throws IllegalArgumentException if the mapper cannot be flattened
+     */
+    List<FieldMapper> flatten(MapperBuilderContext context) {
+        List<FieldMapper> flattenedMappers = new ArrayList<>();
+        ContentPath path = new ContentPath();
+        flatten(context, flattenedMappers, path);
+        return flattenedMappers;
+    }
+
+    private void flatten(MapperBuilderContext context, List<FieldMapper> flattenedMappers, ContentPath path) {
+        assertFlattenable(context, path);
+        path.add(simpleName());
+        for (Mapper mapper : mappers.values()) {
+            if (mapper instanceof FieldMapper fieldMapper) {
+                FieldMapper.Builder fieldBuilder = fieldMapper.getMergeBuilder();
+                fieldBuilder.name = path.pathAsText(mapper.simpleName());
+                flattenedMappers.add(fieldBuilder.build(context));
+            } else if (mapper instanceof ObjectMapper objectMapper) {
+                objectMapper.flatten(context, flattenedMappers, path);
+            }
+        }
+        path.remove();
+    }
+
+    private void assertFlattenable(MapperBuilderContext context, ContentPath path) {
+        if (dynamic != null && context.getDynamic() != dynamic) {
+            throw new IllegalArgumentException(
+                "cannot flatten object ["
+                    + path.pathAsText(simpleName())
+                    + "] because the value of [dynamic] ("
+                    + dynamic
+                    + ") is not compatible with the value from its parent context ("
+                    + context.getDynamic()
+                    + ")"
+            );
+        }
+        if (isEnabled() == false) {
+            throw new IllegalArgumentException(
+                "cannot flatten object [" + path.pathAsText(simpleName()) + "] because the value of [enabled] is [false]"
+            );
+        }
+        if (subobjects.explicit() && subobjects()) {
+            throw new IllegalArgumentException(
+                "cannot flatten object [" + path.pathAsText(simpleName()) + "] because the value of [subobjects] is [true]"
```

Comment:
I wonder if these errors should be more similar to the previous error message e.g. we found an object under an object with subobjects false - we tried to auto-flatten but we couldn't because....